### ISSUE TITLE
Upgraded & downgraded strings behave differently

### ISSUE THREAD

**Issue by FGasper** [perl5-dbi/DBD-mysql#335]
```
> perl -MData::Dumper -MDBI -e'my $dbh = DBI->connect("dbi:mysql:"); $Data::Dumper::Useqq = 1; my $q = q<select "é">; utf8::upgrade($q); print Dumper( $dbh->selectall_arrayref($q) )'
$VAR1 = [
          [
            "\303\203\302\251"
          ]
        ];

> perl -MDBD::mysql -e'print $DBD::mysql::VERSION'
4.050
```
`utf8::upgrade()` merely changes the SV internals and should not affect externally-observable behaviour.

This probably can’t be fixed by default—too much tends to depend on this behaviour—but callers should at least be able to opt out of it.

I propose replacement of `mysql_enable_utf8` and `mysql_enable_utf8mb4` flags with a `mysql_string_mode` parameter that accepts:
- MYSQL_STRING_MODE_PV (default, alas)
- MYSQL_STRING_MODE_UTF8
- MYSQL_STRING_MODE_UTF8MB4
- MYSQL_STRING_MODE_UTF8_BYTES

**Comment by pali** [comment]
Hello!

> Upgraded & downgraded strings behave differently

This is known _feature_ of DBD::mysql and will never change due to existing legacy applications. There was already an attempt to fix this issue but failed as there are many legacy applications.

> I propose replacement of mysql_enable_utf8 and mysql_enable_utf8mb4 flags with a mysql_string_mode parameter that...

This is no-go due to backward compatibility with existing legacy DBD::mysql based applications.

These issues, including broken Unicode support, were fixed in DBD::MariaDB driver, which is fork of DBD::mysql. Therefore if you need working Unicode support and not to deal with these issues, switch from DBD::mysql to DBD::MariaDB driver.

**Comment by FGasper** [comment]
@pali @Grinnz DBD::SQLite fixed the issue by deprecating the old boolean(s), making them passthroughs to the new enum.

As you observe, the default cannot change, but currently there’s no way in this driver to have reliably-working byte-mode strings. It would be nice if that changed, as DBD::MariaDB seems to encompass other changes that may not be a simple drop-in fix.

**Comment by pali** [comment]
> As you observe, the default cannot change, but currently there’s no way in this driver to have reliably-working byte-mode strings.

Exactly! And this was another reason why DBD::MariaDB was created.

> It would be nice if that changed

That is also not possible. We have already tried it in past but due to legacy application and legacy code base it is not possible. And we really spent lot of lot of time on this issue.

> DBD::MariaDB seems to encompass other changes that may not be a simple drop-in fix

DBD::MariaDB _fixed_ or rather _changed_ behavior to be compatible with Unicode support and also important to be compatible with other DBD drivers. So yes, for legacy applications it is not simple drop-in fix.

But I have migrated more applications from DBD::mysql to DBD::MariaDB due to these and Unicode issues, and it was not such a big problem.
